lib/Dialect/StandardOps/Ops.cpp

```diff
@@ -2021,6 +2021,24 @@ OpFoldResult SelectOp::fold(ArrayRef<Attribute> operands) {
   return nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// SignExtendIOp
+//===----------------------------------------------------------------------===//
+
+static LogicalResult verify(SignExtendIOp op) {
+  auto opType = op.getOperand()->getType().dyn_cast_or_null<IntegerType>();
+  if (!opType)
+    return op.emitError("operand type ") << opType << " is not an integer ";
+  auto resType = op.getType().dyn_cast_or_null<IntegerType>();
+  if (!resType)
+    return op.emitError("result type ") << resType << " is not an integer ";
+  if (opType.getWidth() >= resType.getWidth())
+    return op.emitError("result type ")
+           << resType << " must be wider than operand type " << opType;
```

Comment:
I think you need in the verifier:

1) Get the element type:
```
  Type dstEltType = getType();
  if (auto shapedType = dstEltType.dyn_cast<ShapedType>()) 
    dstEltType = shapedType.getElementType();
  Type srcEltType = getType();
  if (auto shapedType = srcEltType.dyn_cast<ShapedType>()) 
    srcEltType = shapedType.getElementType();
```

(I didn't find a helper for this, I think we should have one)

2) Forbid index type for destination (what about source?):

```
   if (dstEltType.isa<IndexType>()) 
      return op.emitError() << "index is not a valid destination type";
```

3) The bit-width comparison you're doing.

Reply:
Understood. The helper I've found is called `getElementTypeOrSelf()` from `IR/TypeUtilities.h`.

For now I'd like to forbid index as both source and destination type. Since the width of index is never stated explicitly, it can be that 

    zexti %0 : index to i64

works on one machine, where the index type is 32 bits wide but it would not when the index type is 64 (or even 128) bits. Treating `zexti` & co dynamically during lowering like `index_cast` (either sign extend or truncate) seems strange to me. I probably would not like that my `zexti` is converted to `llvm.trunc`.

Does this sound reasonable?